### PR TITLE
Update libunwind to latest

### LINKED ISSUE #68640: dotnet crashes on rhel8 arm64 while unwinding stack
The CI job we have that runs the master branch on RHEL8 arm64 is crashing for each test process.

The building of .NET works fine. That is still using the .NET 6.0 sdk, so this is probably a regression.

Interestingly, the CI job that runs on RHEL9 arm64 doesn't have these crashes. One notable difference with RHEL8 which has caused issues with .NET in the past is RHEL8 having a 64k page size rather than the usual 4k.

When running a simple program like:
```
// See https://aka.ms/new-console-template for more information
Console.WriteLine("Hello, World!");
throw new Exception();
```

`gdb /root/repos/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet` gives the following stacktrace which shows a zero pointer being dereferenced.
```
Hello, World!

Thread 1 "dotnet" received signal SIGSEGV, Segmentation fault.
access_mem (as=<optimized out>, addr=0, val=0xffffffffb090, write=<optimized out>, arg=<optimized out>)
    at /root/repos/runtime/src/native/external/libunwind/src/aarch64/Ginit.c:337
337	      *val = *(unw_word_t *) addr;
(gdb) bt
#0  access_mem (as=<optimized out>, addr=0, val=0xffffffffb090, write=<optimized out>, arg=<optimized out>)
    at /root/repos/runtime/src/native/external/libunwind/src/aarch64/Ginit.c:337
#1  0x0000ffffbdfb4728 in is_plt_entry (c=c@entry=0xffffffffb0e0) at /root/repos/runtime/src/native/external/libunwind/src/aarch64/Gstep.c:43
#2  0x0000ffffbdfb4670 in _ULaarch64_step (cursor=0xffffffffb0e0) at /root/repos/runtime/src/native/external/libunwind/src/aarch64/Gstep.c:171
#3  0x0000ffffbdf7aeec in PAL_VirtualUnwind (context=0xffffffffbd80, contextPointers=0xffffffffbce0)
    at /root/repos/runtime/src/coreclr/pal/src/exception/seh-unwind.cpp:566
#4  0x0000ffffbdd8e480 in LazyMachState::unwindLazyState (baseState=<optimized out>, unwoundstate=0xffffffffc160, threadId=<optimized out>, funCallDepth=0, 
    hostCallPreference=<optimized out>) at /root/repos/runtime/src/coreclr/vm/arm64/stubs.cpp:340
#5  0x0000ffffbdbe1cac in HelperMethodFrame::InsureInit (this=0xffffffffdaf8, initialInit=false, unwindState=<optimized out>, 
    hostCallPreference=AllowHostCalls) at /root/repos/runtime/src/coreclr/vm/frames.cpp:1813
#6  0x0000ffffbdbe1c0c in HelperMethodFrame::GetFunction (this=0xffffffffdaf8) at /root/repos/runtime/src/coreclr/vm/frames.cpp:1732
#7  0x0000ffffbdd940fc in ExceptionTracker::InitializeCrawlFrameForExplicitFrame (pcfThisFrame=pcfThisFrame@entry=0xffffffffcd10, 
    pFrame=pFrame@entry=0xffffffffdaf8, pMD=pMD@entry=0xffff4432ff70) at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:1329
#8  0x0000ffffbdd93a58 in ExceptionTracker::ProcessOSExceptionNotification (this=this@entry=0xaaaaaab92a90, 
    pExceptionRecord=pExceptionRecord@entry=0xaaaaaac8dee0, pContextRecord=<optimized out>, pDispatcherContext=pDispatcherContext@entry=0xffffffffd288, 
    dwExceptionFlags=dwExceptionFlags@entry=0, sf=sf@entry=..., pThread=pThread@entry=0xaaaaaab94a90, STState=<optimized out>, 
    pICFSetAsLimitFrame=<optimized out>) at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:1863
#9  0x0000ffffbdd92cd0 in ProcessCLRException (pExceptionRecord=0xaaaaaac8dee0, MemoryStackFp=281474976701536, pContextRecord=0xaaaaaac8db50, 
    pDispatcherContext=pDispatcherContext@entry=0xffffffffd288) at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:1072
#10 0x0000ffffbdd96f6c in UnwindManagedExceptionPass1 (ex=..., frameContext=0xffffffffd6e0) at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:4624
#11 0x0000ffffbdd9725c in DispatchManagedException (ex=..., isHardwareException=<optimized out>)
    at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:4810
#12 0x0000ffffbdcfc188 in IL_Throw (obj=<optimized out>) at /root/repos/runtime/src/coreclr/vm/jithelpers.cpp:4024
#13 0x0000ffff445aa6a8 in ?? ()
#14 0x0000ffbf3680c858 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

@janvorli do you know what could be the cause of this regression?

### PR DESCRIPTION
**Pull request by janvorli**
This change updates libunwind copy by a fix for non-4kB page sizes
that is needed e.g. for Linux arm64 with kernel set to use 64kB page
sizes (RHEL8 is one such case).

It also applies a fix for incorrect register storage in ARM64
getcontext.

Close #68640

### CODE REVIEW

File: src/native/external/libunwind-version.txt [hunk 1]
```diff
@@ -1,10 +1,6 @@
-v1.6.2
-https://github.com/libunwind/libunwind/commit/b3ca1b59a795a617877c01fe5d299ab7a07ff29d
+main@e07b43c02d5cf1ea060c018fdf2e2ad34b7c7d80
+https://github.com/libunwind/libunwind/commit/e07b43c02d5cf1ea060c018fdf2e2ad34b7c7d80
```
Comment (am11) [inline src/native/external/libunwind-version.txt:2]: This has pulled up a lot of unnecessary changes. An alternative fix is to keep pointing to released version 1.6.2, and apply patch from PR 330, as we have done for the entire loongarch64 support.

**Comment by am11** [comment]
> for Linux arm64 with kernel set to use 64kB page sizes (RHEL8 is one such case).

Apparently, R2R was also broken on such linux-arm64 system: https://github.com/dotnet/runtime/pull/68845#discussion_r864355088. We do not test against non-4K-page configuration in CI, so there could be more stuff broken besides libunwind.

File: src/native/external/libunwind-version.txt [hunk 2]
```diff
@@ -1,10 +1,6 @@
...
 
 Reapply changes from https://github.com/dotnet/runtime/commit/1b5719c2e3dde393531eaeb5b5cde05dabeef5b8
-Apply https://github.com/libunwind/libunwind/pull/317
 Apply https://github.com/libunwind/libunwind/pull/333
 Apply https://github.com/libunwind/libunwind/pull/342
```
[the remark this answers is not in the document]
Comment (janvorli) [reply src/native/external/libunwind-version.txt:6]: I was trying to be conservative and include only the merged in changes. The PR 353 is an obvious bug fix though, so I agree that it would make sense to apply that one too. 
The other two are related to walking over a signal frame and we never do that, we unwind that differently in PAL_VirtualUnwind.

**Comment by janvorli** [comment]
> We do not test against non-4K-page configuration in CI, so there could be more stuff broken besides libunwind.

All coreclr pri1 tests are passing on my local system with this change.

File: src/native/external/libunwind-version.txt (hunk 1 above)
Comment (janvorli) [reply src/native/external/libunwind-version.txt:2]: The changes that it has pulled  in all look ok and simple. Besides the loongarch64 changes that are now included in the main and the PR 330, there were only five PRs - one just a licence header update, one adding couple of ifdefs for freebsd, one fixing malloc issue, one fixing a race and one updating the latex doc.

**Comment by am11** [comment]
> The windows builds that pull in libunwind are failing as they don't know the sysconf that is now being called to get the page size. I need to fix it.

Ah probably need to add it `libunwind/src/win/pal-single-threaded.c` (we have `getpagesize` also defined there).

File: src/native/external/libunwind-version.txt (hunk 1 above)
Comment (am11) [reply src/native/external/libunwind-version.txt:2]: Loongarch and  other patches were applied on top of 1.6.2 without resetting to libunwind's master branch, which is still using version 1.6.1-rc1. My comment was we can apply PR 330 in similar way which fixes the issue.

If you prefer this approach, ok. I think it just makes the versioning a bit ambiguous, but not a big deal.

**Comment by janvorli** [comment]
Sigh, there is another problem, this time due to the PR 327. The change has some invalid C standard code that Unix compilers compile because of an extended behavior, but VC does not:
```c
  const unsigned long full_path_buff_sz = 1024;
  char                full_path_buff[full_path_buff_sz];
```
`const` doesn't mean a constant, but rather a "read only" variable in C and VC doesn't support declaring arrays with non-constant size.

**Comment by janvorli** [comment]
So I'll probably just follow your advice and pull in the two PRs (330 and 353)

**Comment by janvorli** [comment]
@am11 I've changed the PR to just add the PR 330 and 353 and add the windows getconf implementation.

**Review by am11 — COMMENTED** [review]
Thank you. I will submit your patch along with https://github.com/am11/libunwind/commit/c59e547193e1ab05360dcd1a1ae71c4b2c6e4845 upstream to make future updates easier.

Update: https://github.com/libunwind/libunwind/pull/358

**Comment by janvorli** [comment]
The Windows arm legs are failing
```
2022-05-05T15:50:38.8303558Z init.c.obj : error LNK2019: unresolved external symbol _write@12 referenced in function _print_error@4
```
It seems the print error has pulled in "write" function that we also need to implement for Windows.

File: src/native/external/libunwind-version.txt [hunk 3]
```diff
@@ -3,8 +3,10 @@ https://github.com/libunwind/libunwind/commit/b3ca1b59a795a617877c01fe5d299ab7a0
 
 Reapply changes from https://github.com/dotnet/runtime/commit/1b5719c2e3dde393531eaeb5b5cde05dabeef5b8
 Apply https://github.com/libunwind/libunwind/pull/317
+Apply https://github.com/libunwind/libunwind/pull/330
 Apply https://github.com/libunwind/libunwind/pull/333
 Apply https://github.com/libunwind/libunwind/pull/342
+Apply https://github.com/libunwind/libunwind/pull/353
```
Comment (am11) [inline src/native/external/libunwind-version.txt:9]:
```suggestion
Apply https://github.com/libunwind/libunwind/pull/353
Apply https://github.com/libunwind/libunwind/pull/358
```

I have separated unrelated change in 359.

**Review by am11 — APPROVED** [review]
LGTM, thank you.

**Review by jkotas — APPROVED** [review]
